### PR TITLE
Add logging with state to Bayesian Optimizer

### PR DESCRIPTION
**Pull request by adrianmolzon**
Hey all, long time listener first time contributor. I've created two new methods for the BayesianOptimizer class to allow users to save and load their optimizers while maintaining state. I've added a new section to the basic tour to demonstrate this cleaner method for saving and loading their optimizer.

A couple notes, users using a custom acquisition function will have to write their own get_acquisition_params and set_acquisition_params class functions in order to preserve exact state, otherwise there might be a loss of state when saving and loading. This should still be strictly better than the previous flow for saving and loading optimizers which ignored the acquisition parameters.

If this pull request gets merged into master, I'd like to remove the previous saving methodology, but I'll save that for a future PR.

### CODE REVIEW

**Comment by adrianmolzon** [comment]
@till-m

**Comment by till-m** [comment]
@adrianmolzon I realized the coverage bot is broken, so let me fix that first before moving on with this PR

File: bayes_opt/acquisition.py [hunk 1]
```diff
@@ -69,18 +69,57 @@ def __init__(self, random_state: int | RandomState | None = None) -> None:
...
+
+    def set_acquisition_params(self, params: dict) -> None:
+        """Set the acquisition function parameters.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary containing the acquisition function parameters.
+        """
+        return {}
```
Comment (till-m) [inline bayes_opt/acquisition.py:122]: While I don't think that we should enforce that these methods are defined for subclasses (i.e. we don't need to make them `abstractmethod`s), we should make them raise a `NotImplementedError` so that if people try to store/load with custom acqs without having custom, properly defined `get`/`set` methods, it will error out instead of letting them believe that it just magically knows which parameters to store and load

File: bayes_opt/acquisition.py [hunk 2]
```diff
@@ -129,6 +168,15 @@ def suggest(
...
 
+    def _fit_gp(self, gp: GaussianProcessRegressor, target_space: TargetSpace) -> None:
+        # Sklearn's GP throws a large number of warnings at times, but
+        # we don't really need to see them here.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            gp.fit(target_space.params, target_space.target)
+            if target_space.constraint is not None:
+                target_space.constraint.fit(target_space.params, target_space._constraint_values)
+
```
Comment (till-m) [inline bayes_opt/acquisition.py:179]: I think you moved this function declaration. Any specific reason why?

File: bayes_opt/acquisition.py [hunk 3]
```diff
@@ -1035,3 +1199,42 @@ def suggest(
...
+        dict
+            Dictionary containing the current acquisition function parameters.
+        """
+        return {
+            "base_acquisitions_params": [acq.get_acquisition_params() for acq in self.base_acquisitions],
+            "gains": self.gains.tolist(),
+            "previous_candidates": self.previous_candidates.tolist()
+            if self.previous_candidates is not None
+            else None,
+            "random_states": [acq._serialize_random_state() for acq in self.base_acquisitions]
```
Comment (till-m) [inline bayes_opt/acquisition.py:1217]: Are the random states for the base_acqs not already stored in Line 1212 under `"base_acquisition_params"`?

File: examples/optimizer_state.json
[the remark this answers is not in the document]
Comment (till-m) [reply examples/optimizer_state.json:1]: The same holds for `logs.log`.

File: bayes_opt/acquisition.py (hunk 3 above)
Comment (adrianmolzon) [reply bayes_opt/acquisition.py:1217]: Each of these base acquisition functions in base_acquisitions have:

1. A random state initialized inside of their acquisition function
2. A high level random state from the BayesianOptimizer

The base_acquisition_params key holds the acquisition function parameters, which will include inside it 1. The random_states key holds the random state of the BayesianOptimizer random state, 2.

File: examples/optimizer_state.json
[the remark this answers is not in the document]
Comment (adrianmolzon) [reply examples/optimizer_state.json:1]: removed

File: bayes_opt/acquisition.py [hunk 4]
```diff
@@ -100,26 +100,42 @@ def _deserialize_random_state(self, state_dict: dict | None) -> None:
...
+    def _fit_gp(self, gp: GaussianProcessRegressor, target_space: TargetSpace) -> None:
+        # Sklearn's GP throws a large number of warnings at times, but
+        # we don't really need to see them here.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            gp.fit(target_space.params, target_space.target)
+            if target_space.constraint is not None:
+                target_space.constraint.fit(target_space.params, target_space._constraint_values)
+
+    def get_acquisition_params(self):
```
Comment (till-m) [inline bayes_opt/acquisition.py:112]: I think the return type of this should be `dict[str, Any] | NoReturn`

**Comment by till-m** [comment]
I think this is ready to go, I will merge this in the coming days if there are no objections.

**Comment by till-m** [comment]
Thanks for your contribution!
